go-controller/hybrid-overlay/pkg/controller/node_windows.go

```diff
@@ -0,0 +1,375 @@
+	// while the base network remains across reboots.
+
+	const (
+		baseNetworkName = "BaseOVNKubernetesHybridOverlayNetwork"
+		fakeSubnetVNI   = types.HybridOverlayVNI + 1
+	)
+
+	// Unused subnet and gateway IP required to create the base overlay network.
+	// This subnet is actually invisible to the PODs/Nodes.
+	_, fakeSubnetCIDR, _ := net.ParseCIDR("100.64.0.0/30")
+	fakeSubnetGateway := net.ParseIP("100.64.0.2")
+
+	baseNetwork := GetExistingNetwork(baseNetworkName, fakeSubnetCIDR.String(), fakeSubnetGateway.String())
+	if baseNetwork != nil {
+		// nothing to do
+		return nil
+	}
+
+	baseNetworkInfo := NetworkInfo{
+		AutomaticDNS: false,
+		IsPersistent: true,
+		Name:         baseNetworkName,
+		Subnets: []SubnetInfo{{
+			AddressPrefix:  fakeSubnetCIDR,
+			GatewayAddress: fakeSubnetGateway,
+			VSID:           fakeSubnetVNI,
+		}},
+	}
+
+	klog.Infof("Creating the base overlay network '%s'.", baseNetworkName)
+
+	// Retrieve the network schema object
+	baseNetworkSchema, err := baseNetworkInfo.GetHostComputeNetworkConfig()
+	if err != nil {
+		return fmt.Errorf("Unable to generate a schema to create a base overlay network, error: %v", err)
+	}
+
+	klog.Infof("Base network creation may take up to a minute to complete...")
+	// Create the base network from schema object
+	if _, err = baseNetworkSchema.Create(); err != nil {
+		return fmt.Errorf("Unable to create the base overlay network, error: %v", err)
+	}
+
+	// Workaround for a limitation in the Windows HNS service. We need
+	// to manually duplicate persistent routes that used to be on the
+	// physical network interface to the newly created host vNIC
+	if err = DuplicatePersistentIPRoutes(); err != nil {
+		return fmt.Errorf("Unable to refresh the persistent IP routes, error: %v", err)
+	}
+
+	return nil
+}
+
+// Start is the top level function to run hybrid-sdn in node mode
+func (n *NodeController) Start(wf *factory.WatchFactory) error {
+	return houtil.StartNodeWatch(n, wf)
+}
+
+// Add sets up VXLAN tunnels to other nodes
+// For a windows node, this means watching for all nodes and programming the routing
+func (n *NodeController) Add(node *kapi.Node) {
+	if node.Status.NodeInfo.MachineID == n.machineID {
+		// Initialize or the local node (or reconfigure it if the addresses
+		// have changed) by creating the network object and setting up
+		// all the VXLAN tunnels towards other nodes
+		cidr, nodeIP := getNodeSubnetAndIP(node)
+		if (cidr != nil && !houtil.SameIPNet(cidr, n.localNodeCIDR)) || (nodeIP != nil && nodeIP.Equal(n.localNodeIP)) {
+			n.localNodeCIDR = cidr
+			n.localNodeIP = nodeIP
+			if err := n.initSelf(node, cidr); err != nil {
+				klog.Errorf("failed to initialize node: %v", err)
+			}
+		}
+		return
+	}
+
+	cidr, nodeIP, drMAC, err := getNodeDetails(node)
+	if cidr == nil || nodeIP == nil || drMAC == nil {
+		klog.V(5).Infof("cleaning up hybrid overlay resources for node %q because: %v", node.Name, err)
```

Comment:
so if we cant get node details, it's ok for it just to be a debug log? or should we be issuing an error log here?

Reply:
@trozet the node might not have all the annotations when it's added to the API, since the annotations get added by the ovnkube --node process itself (DRMAC) and the ovnkube master process (node subnet). It's not an error if they don't have the annotations. But if they don't for whatever reason, or if the subnet changes or something, we should clean up any resources the node had locally.